File: tokenizer.c
#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>
#include <string.h> 
#include "grammar.h"

#define MAX_TOKEN_LENGTH 32

typedef struct Token {
  TokenType type;
  char value[MAX_TOKEN_LENGTH + 1];
} Token;

int is_keyword(const char* str) {
  return strcmp(str, KEYWORD_INPUT) == 0 || strcmp(str, KEYWORD_OUTPUT) == 0;
}

TokenType get_token_type(const char* str) {
  if (isdigit(*str)) {
    return TOKEN_NUMBER;
  } else if (isalpha(*str)) {
    return TOKEN_IDENTIFIER;
  } else if (is_keyword(str)) {
    return TOKEN_KEYWORD;
  } else if (*str == '+') {
    return TOKEN_PLUS;
  } else if (*str == '-') {
    return TOKEN_MINUS;
  } else if (*str == '*') {
    return TOKEN_MULTIPLY;
  } else if (*str == '/') {
    return TOKEN_DIVIDE;
  } else if (*str == '(') {
    return TOKEN_LPAREN;
  } else if (*str == ')') {
    return TOKEN_RPAREN;
  } else if (*str == ';') {
    return TOKEN_SEMICOLON;
  } else {
    return TOKEN_ERROR;
  }
}

Token* tokenize(const char* input_string) {
  int num_tokens = 0;
  int max_tokens = 10;
  Token* tokens = malloc(sizeof(Token) * max_tokens);

  const char* current_char = input_string;
  while (*current_char != '\0') {
    // Skip whitespace
    while (isspace(*current_char)) {
      current_char++;
    }

    if (*current_char == '\0') {
      break;
    }

    int token_length = 0;
    while (*current_char != '\0' && !isspace(*current_char) && token_length < MAX_TOKEN_LENGTH) {
      if (*current_char == '+' || *current_char == '-' || *current_char == '*' || *current_char == '/' ||
          *current_char == '(' || *current_char == ')' || *current_char == ';') {
        if (token_length > 0) {
          // If we have characters accumulated in the token buffer, create a token with them first
          tokens[num_tokens].value[token_length] = '\0'; // Null terminate
          tokens[num_tokens].type = get_token_type(tokens[num_tokens].value);
          num_tokens++;
          token_length = 0;
          if (num_tokens == max_tokens) {
            max_tokens *= 2;
            tokens = realloc(tokens, sizeof(Token) * max_tokens);
          }
        }
        // Create a token for the operator or delimiter
        tokens[num_tokens].value[0] = *current_char;
        tokens[num_tokens].value[1] = '\0'; // Null terminate
        tokens[num_tokens].type = get_token_type(tokens[num_tokens].value);
        num_tokens++;
        if (num_tokens == max_tokens) {
          max_tokens *= 2;
          tokens = realloc(tokens, sizeof(Token) * max_tokens);
        }
        current_char++;
      } else {
        // Accumulate characters into the token buffer
        tokens[num_tokens].value[token_length++] = *current_char++;
      }
    }

    if (token_length > 0) {
      // If we have characters accumulated in the token buffer, create a token with them
      tokens[num_tokens].value[token_length] = '\0'; // Null terminate
      tokens[num_tokens].type = get_token_type(tokens[num_tokens].value);
      num_tokens++;
      if (num_tokens == max_tokens) {
        max_tokens *= 2;
        tokens = realloc(tokens, sizeof(Token) * max_tokens);
      }
    }
  }

  // Resize tokens array to actual number of tokens
  tokens = realloc(tokens, sizeof(Token) * (num_tokens + 1));
  tokens[num_tokens].type = TOKEN_ERROR; 
  // Mark end of tokens

  return tokens;
}

// Example usage (can be removed)
int main() { 
  char input_string[] = "INPUT x; OUTPUT x + 2 * (y - 3);";
  Token* tokens = tokenize(input_string);

  int i = 0;
  while (tokens[i].type != TOKEN_ERROR) {
    printf("Token %d: type=%d, value=%s\n", i + 1, tokens[i].type, tokens[i].value);
    i++;
  }

  free(tokens);
  return 0;
}

File: tokenizer.py
from enum import Enum

class TokenType(Enum):
    IDENTIFIER = 0
    KEYWORD = 1
    PLUS = 2
    MINUS = 3
    MULTIPLY = 4
    DIVIDE = 5
    LPAREN = 6
    RPAREN = 7
    SEMICOLON = 8
    NUMBER = 9
    ERROR = 10

class TokenizationError(Exception):
    def __init__(self, message, position):
        super().__init__(message)
        self.position = position

def is_keyword(string):
    """Checks if a string matches one of the reserved keywords."""
    return string in ("INPUT", "OUTPUT")

def get_token_type(char):
    """Determines the token type based on the first character."""
    if char.isdigit():
        return TokenType.NUMBER
    elif char.isalpha():
        return TokenType.IDENTIFIER
    elif char == "+":
        return TokenType.PLUS
    elif char == "-":
        return TokenType.MINUS
    elif char == "*":
        return TokenType.MULTIPLY
    elif char == "/":
        return TokenType.DIVIDE
    elif char == "(":
        return TokenType.LPAREN
    elif char == ")":
        return TokenType.RPAREN
    elif char == ";":
        return TokenType.SEMICOLON
    else:
        return TokenType.ERROR

def tokenize(input_string):
    """Tokenizes the input string into a list of tokens and errors."""
    tokens = []
    errors = []
    current_char = 0
    while current_char < len(input_string):
        # Skip whitespace
        while current_char < len(input_string) and input_string[current_char].isspace():
            current_char += 1

        if current_char == len(input_string):
            break

        token_value = ""
        # Check for unary minus or negative number
        if input_string[current_char] == "-":
            next_char = current_char + 1
            if next_char < len(input_string) and input_string[next_char].isdigit():
                # Negative number
                token_value = "-"
                current_char = next_char
                while current_char < len(input_string) and input_string[current_char].isdigit():
                    token_value += input_string[current_char]
                    current_char += 1
            else:
                # Unary minus
                token_value = "-"
                current_char += 1
        elif input_string[current_char] in ('+', '*', '/', '(', ')', ';'):
            # Single-character operators and symbols
            token_value = input_string[current_char]
            current_char += 1
        else:
            # Other characters
            while current_char < len(input_string) and not input_string[current_char].isspace() \
                    and input_string[current_char] not in ('+', '-', '*', '/', '(', ')', ';'):
                token_value += input_string[current_char]
                current_char += 1

        if token_value:
            token_type = get_token_type(token_value[0])
            if token_type == TokenType.ERROR and len(token_value) > 1:
                # Handle potential multi-character errors (e.g., unexpected symbols)
                errors.append(TokenizationError(f"Unexpected character(s): {token_value}", current_char - 1))
            else:
                tokens.append((token_type, token_value))

    # Add sentinel token
    tokens.append((TokenType.ERROR, None))
    return tokens, errors

# Example usage
input_string = "INPUT x; OUTPUT x + 2 * (y - 3);"
tokens, errors = tokenize(input_string)

if errors:
    for error in errors:
        print(f"Error at position {error.position}: {error}")
else:
    print("Tokenization successful!")
    for token_type, value in tokens:
        print(f"Token: type={token_type.name}, value={value}")
